### PR TITLE
Sg2/various perf improvements

### PR DESCRIPTION
**Pull request by elizabeth-legros**
- Many changes to ElementID to make more performant
- Change MetadataCollection to ensure only serialized layer metadata blocks are saved out
- Small change to layer iteration to avoid enumerator cost
- Cache off the list of context nodes for getting context nodes in order
- Changes to call ReconcretizeAll fewer times (still can maybe reduce)
- Fixing bug in our importers that caused multiple import calls per graph

### CODE REVIEW

**Review by bencloward — APPROVED** [review]

File: com.unity.sg2/Editor/ContextLayeredDataStorage/ElementID.cs [hunk 1]
```diff
@@ -23,71 +21,281 @@ public int GetHashCode(ElementID obj)
...
+                m_hash = GetDeterministicStringHash(m_fullPath, 0, m_fullPath.Length);
+            }
+            return m_hash;
+        }
+
+        private static int GetDeterministicStringHash(char[] str, int startIndex, int length)
+        {
+            unchecked
+            {
+                int hash1 = (5381 << 16) + 5381;
```
[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.sg2/Editor/ContextLayeredDataStorage/ElementID.cs:237]: this is some john carmack fast inverse square root shit, and I love it

Comment (Nightmask3) [reply com.unity.sg2/Editor/ContextLayeredDataStorage/ElementID.cs:237]: is there any reason why you're using these specific values??

File: com.unity.sg2/Editor/ContextLayeredDataStorage/MetadataBlock.cs [hunk 2]
```diff
@@ -106,13 +108,17 @@ public void OnAfterDeserialize()
         public void OnBeforeSerialize()
         {
             serializedBlocks = new List<SerializedBlock>();
-            foreach(var (key, value) in this)
+            foreach (var (key, value) in this)
             {
-                serializedBlocks.Add(new SerializedBlock()
+                if (owner.GetHierarchyValue(owner.Search(key).Element) > 0)
```
[the remark this answers is not in the document]
Comment (elizabeth-legros) [reply com.unity.sg2/Editor/ContextLayeredDataStorage/MetadataBlock.cs:113]: Cause we were serializing data that we werent supposed to be.

File: com.unity.sg2/Editor/ContextLayeredDataStorage/ElementID.cs [hunk 3]
```diff
@@ -23,71 +21,281 @@ public int GetHashCode(ElementID obj)
     }
 
     [Serializable]
-    public struct ElementID
+    public class ElementID : ISerializationCallbackReceiver
```
[the remark this answers is not in the document]
Comment (elizabeth-legros) [reply com.unity.sg2/Editor/ContextLayeredDataStorage/ElementID.cs:24]: largely this was just to handle the issue of changing some of the underlying serialized data and so to handle changing to the new form

**Review by Nightmask3 — APPROVED** [review]

**Pull request merged**